tests/test_graphml.py

```diff
@@ -0,0 +1,399 @@
+
+    def test_string(self):
+        graph_xml = self.HEADER.format(
+            """
+            <key id="d0" for="node" attr.name="test" attr.type="string">
+            <default>yellow</default>
+            </key>
+            <graph id="G" edgedefault="directed">
+            <node id="n0">
+                <data key="d0">blue</data>
+            </node>
+            <node id="n1"/>
+            <node id="n2">
+                <data key="d0">yellow</data>
+            </node>
+            </graph>
+            """
+        )
+
+        with tempfile.NamedTemporaryFile("wt") as fd:
+            fd.write(graph_xml)
+            fd.flush()
+            graphml = retworkx.read_graphml(fd.name)
+            graph = graphml[0]
+            nodes = [
+                {"id": "n0", "test": "blue"},
+                {"id": "n1", "test": "yellow"},
+                {"id": "n2", "test": "yellow"},
+            ]
+            edges = []
+            self.assertGraphEqual(graph, nodes, edges, directed=True)
+
+    def test_convert_error(self):
+        graph_xml = self.HEADER.format(
+            """
+            <key id="d0" for="node" attr.name="test" attr.type="{}">
+            <default>blah</default>
+            </key>
+            """
+        )
+
+        for type in ["boolean", "int", "float", "double"]:
+            self.assertGraphMLRaises(graph_xml=graph_xml.format(type))
+
+    def test_invalid_xml(self):
+        graph_xml = self.HEADER.format(
+            """
+            <key id="d0" for="node" attr.name="test" attr.type="string">
+            </default>
+            </key>
+            """
+        )
+        self.assertGraphMLRaises(graph_xml)
+
+    def test_invalid_edgedefault(self):
+        graph_xml = self.HEADER.format(
+            """
+            <graph id="G" edgedefault="UnDir">
+            <node id="n0"/>
+            </graph>
+            """
+        )
+        self.assertGraphMLRaises(graph_xml)
+
+    def test_missing_node_id(self):
+        graph_xml = self.HEADER.format(
+            """
+            <graph id="G" edgedefault="directed">
+            <node/>
+            </graph>
+            """
+        )
+        self.assertGraphMLRaises(graph_xml)
+
+    def test_missing_key_for_node(self):
+        graph_xml = self.HEADER.format(
+            """
+            <key id="d0" for="edge" attr.name="color" attr.type="string"/>
+            <graph id="G" edgedefault="directed">
+            <node id="n0">
+                <data key="d0">blue</data>
+            </node>
+            </graph>
+            """
+        )
+        self.assertGraphMLRaises(graph_xml)
+
+    def test_invalid_key_type(self):
+        graph_xml = self.HEADER.format(
+            """
+            <key id="d0" for="node" attr.name="test" attr.type="List[int]"/>
+            """
+        )
+        self.assertGraphMLRaises(graph_xml)
+
+    def test_unsupported_key_domain(self):
```

Comment:
In mixed graphs we are currently raising an error but I decided to ignore the definition of hyperedges and ports. Do you think it's better to raise an error in these cases too?

Reply:
Yeah, I think it would probably be better to error than ignore in those cases. I think it would be more frustrating to a user that is trying to load a file with hyperedges or ports if those were just missing from the output graph object. An error will at least be explicit and give the response that the input file can't be represented in retworkx.

Reply:
Done in e49eb76.